### ISSUE TITLE
Race condition for PV callback unregistration with the ca thread.

### ISSUE THREAD

**Issue by mccullerlp** [pyepics/pyepics#33]
Hi, good to see there is still active development on this project.

Isee very rare exceptions coming from the ca thread due to a rare race condition between unregistering callbacks in other threads and calling all of the callbacks in a ca-thread loop:

``` python
    def run_callbacks(self):
        """run all user-defined callbacks with the current data

        Normally, this is to be run automatically on event, but
        it is provided here as a separate function for testing
        purposes.
        """
        for index in sorted(list(self.callbacks.keys())):
            self.run_callback(index)
```

Occasionally in this ca-thread loop, the index can have been unregistered from the program thread. Either a lock needs to be put around accesses to self.callbacks, or since most operations are atomic-enough, you can actually fix this with

``` python
    def run_callback(self, index):
        """run a specific user-defined callback, specified by index,
        with the current data
        Note that callback functions are called with keyword/val
        arguments including:
             self._args  (all PV data available, keys = __fields)
             keyword args included in add_callback()
             keyword 'cb_info' = (index, self)
        where the 'cb_info' is provided as a hook so that a callback
        function  that fails may de-register itself (for example, if
        a GUI resource is no longer available).
        """
        try:
            fcn, kwargs = self.callbacks[index]
        except KeyError:
            #key suddenly stopped existing, must have been unregistered
            return
        kwd = copy.copy(self._args)
        kwd.update(kwargs)
        kwd['cb_info'] = (index, self)
        if hasattr(fcn, '__call__'):
            fcn(**kwd)
```

For personal interest, I'm actually using this library through UChicago physics. I should come say Hi sometime as it looks like the main development is though the university?

**Comment by newville** [comment]
Hi @mccullerlp:  Thanks, that's a good point.  I've never seen that problem myself, but I also tend to not make a lot of threads with PVs with lots of callbacks.  I can see that this could KeyError could occur.  I'll post a fix.

That's great that you're at U Chicago Physics!   I'd be interested in what project you are using Epics with, but maybe it's more appropriate if you send me a separate email.

**Comment by klauer** [comment]
This was fixed as suggested by @mccullerlp in c8902a884e5f651d6d14fff5e089dd36b9bc4463.
